vote/src/vote_state_view/field_frames.rs

```diff
@@ -0,0 +1,361 @@
+        static_assertions::const_assert!(core::mem::align_of::<EpochCreditsItem>() == 1);
+    };
+
+    fn len(&self) -> usize {
+        self.len as usize
+    }
+}
+
+impl EpochCreditsItem {
+    #[inline]
+    pub fn epoch(&self) -> u64 {
+        u64::from_le_bytes(self.epoch)
+    }
+    #[inline]
+    pub fn credits(&self) -> u64 {
+        u64::from_le_bytes(self.credits)
+    }
+    #[inline]
+    pub fn prev_credits(&self) -> u64 {
+        u64::from_le_bytes(self.prev_credits)
+    }
+}
+
+impl From<&EpochCreditsItem> for (Epoch, u64, u64) {
+    fn from(item: &EpochCreditsItem) -> Self {
+        (item.epoch(), item.credits(), item.prev_credits())
+    }
+}
+
+pub(super) struct RootSlotView<'a> {
+    frame: RootSlotFrame,
+    buffer: &'a [u8],
+}
+
+impl<'a> RootSlotView<'a> {
+    pub(super) fn new(frame: RootSlotFrame, buffer: &'a [u8]) -> Self {
+        Self { frame, buffer }
+    }
+}
+
+impl RootSlotView<'_> {
+    pub(super) fn root_slot(&self) -> Option<Slot> {
+        if !self.frame.has_root_slot {
+            None
+        } else {
+            let root_slot = {
+                let mut cursor = std::io::Cursor::new(self.buffer);
+                cursor.consume(1);
+                solana_serialize_utils::cursor::read_u64(&mut cursor).unwrap()
+            };
+            Some(root_slot)
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample))]
+pub(super) struct RootSlotFrame {
+    has_root_slot: bool,
+}
+
+impl RootSlotFrame {
+    pub(super) fn total_size(&self) -> usize {
+        1 + self.size()
+    }
+
+    pub(super) fn size(&self) -> usize {
+        if self.has_root_slot {
+            core::mem::size_of::<Slot>()
+        } else {
+            0
+        }
+    }
+
+    pub(super) fn read(cursor: &mut std::io::Cursor<&[u8]>) -> Result<Self> {
+        let has_root_slot = solana_serialize_utils::cursor::read_bool(cursor)
+            .map_err(|_err| VoteStateViewError::ParseError)?;
+        let frame = Self { has_root_slot };
+        cursor.consume(frame.size());
+        Ok(frame)
+    }
+}
+
+pub(super) struct PriorVotersFrame;
+
+impl ListFrame for PriorVotersFrame {
+    type Item = PriorVotersItem;
+
+    #[cfg(test)]
+    const ASSERT_ITEM_ALIGNMENT: () = {
+        static_assertions::const_assert!(core::mem::align_of::<PriorVotersItem>() == 1);
+    };
+
+    fn len(&self) -> usize {
+        const MAX_ITEMS: usize = 32;
+        MAX_ITEMS
+    }
+
+    fn total_size(&self) -> usize {
+        self.total_item_size() +
+            core::mem::size_of::<u64>() /* idx */ +
+            core::mem::size_of::<bool>() /* is_empty */
+    }
+}
+
+#[repr(C)]
+pub(super) struct PriorVotersItem {
+    voter: Pubkey,
+    start_epoch_inclusive: [u8; 8],
+    end_epoch_exclusive: [u8; 8],
+}
+
+impl PriorVotersFrame {
+    pub(super) fn read(cursor: &mut std::io::Cursor<&[u8]>) {
+        cursor.consume(PriorVotersFrame.total_size());
+    }
+}
```

Comment:
no tests for the individual frame parsing / reading?

Reply:
Yeah I don't think we need it. It could be garbage and the runtime doesn't care because it never reads it. I reverted the list frame changes for prior voters and just made it implement `total_size`. Let me know if you prefer a different approach